File: scripts/run_migrations.py
#!/usr/bin/env python3
"""
MasseurMatch Migration Runner - Standalone Version
Executes all SQL migrations using HTTP API directly
"""

import os
import sys
import glob
import json
from pathlib import Path
import urllib.request
import urllib.error
import time

class MigrationRunner:
    def __init__(self, supabase_url: str, supabase_key: str):
        """Initialize with Supabase credentials"""
        self.supabase_url = supabase_url.rstrip('/')
        self.supabase_key = supabase_key
        self.migrations_dir = Path(__file__).parent.parent / "supabase" / "migrations"
        self.executed = 0
        self.skipped = 0
        self.errors = 0
        self.failed_migrations = []

    def get_migration_files(self):
        """Get all migration files in correct order"""
        migrations = sorted(glob.glob(str(self.migrations_dir / "*.sql")))
        return migrations

    def read_migration(self, filepath):
        """Read migration SQL file"""
        with open(filepath, 'r', encoding='utf-8') as f:
            return f.read()

    def execute_migration(self, filepath):
        """Execute a single migration file using Supabase REST API"""
        filename = Path(filepath).name
        
        try:
            sql_content = self.read_migration(filepath)
            
            if not sql_content.strip():
                print(f"⏭️  Skipping {filename} (empty file)")
                self.skipped += 1
                return True
            
            # Execute via Supabase REST API
            url = f"{self.supabase_url}/rest/v1/rpc/exec_sql"
            headers = {
                'Authorization': f'Bearer {self.supabase_key}',
                'Content-Type': 'application/json',
                'apikey': self.supabase_key
            }
            
            # Split into individual statements
            statements = [s.strip() for s in sql_content.split(';') if s.strip()]
            
            for stmt in statements:
                if not stmt.strip():
                    continue
                    
                data = json.dumps({'sql': stmt}).encode('utf-8')
                req = urllib.request.Request(url, data=data, headers=headers, method='POST')
                
                try:
                    with urllib.request.urlopen(req, timeout=30) as response:
                        result = response.read().decode('utf-8')
                        if response.status not in [200, 201]:
                            print(f"⚠️  Warning in {filename}: {result}")
                except urllib.error.HTTPError as e:
                    error_msg = e.read().decode('utf-8')
                    # Ignore "already exists" errors
                    if 'already exists' in error_msg.lower() or 'duplicate' in error_msg.lower():
                        print(f"✓ {filename} (already exists - skipped)")
                        self.skipped += 1
                        return True
                    else:
                        print(f"❌ Error in {filename}: {error_msg}")
                        self.errors += 1
                        self.failed_migrations.append(filename)
                        return False
            
            print(f"✓ {filename}")
            self.executed += 1
            return True
            
        except Exception as e:
            print(f"❌ Failed {filename}: {str(e)}")
            self.errors += 1
            self.failed_migrations.append(filename)
            return False

    def run_all(self):
        """Execute all migrations"""
        migrations = self.get_migration_files()
        
        if not migrations:
            print("❌ No migration files found!")
            return False
        
        print(f"Found {len(migrations)} migrations")
        print("=" * 60)
        
        for filepath in migrations:
            self.execute_migration(filepath)
            time.sleep(0.5)  # Rate limiting
        
        print("=" * 60)
        print(f"\n✓ Executed: {self.executed}")
        print(f"⏭️  Skipped: {self.skipped}")
        print(f"❌ Errors: {self.errors}")
        
        if self.failed_migrations:
            print(f"\nFailed migrations:")
            for m in self.failed_migrations:
                print(f"  - {m}")
        
        return self.errors == 0

def main():
    # Get credentials from environment
    supabase_url = os.getenv('SUPABASE_URL')
    supabase_key = os.getenv('SUPABASE_SERVICE_ROLE_KEY')
    
    if not supabase_url or not supabase_key:
        print("❌ Missing environment variables!")
        print("Set SUPABASE_URL and SUPABASE_SERVICE_ROLE_KEY")
        sys.exit(1)
    
    runner = MigrationRunner(supabase_url, supabase_key)
    success = runner.run_all()
    
    sys.exit(0 if success else 1)

if __name__ == '__main__':
    main()

File: scripts/run-migrations.js
#!/usr/bin/env node

/**
 * Run Database Migrations
 * Calls the migration API endpoint to execute all SQL migrations
 */

const DEV_PORT = process.env.DEV_PORT || process.env.NEXT_PUBLIC_DEV_PORT || '5000';
const baseUrl = process.env.VERCEL_URL
  ? `https://${process.env.VERCEL_URL}`
  : (process.env.NEXT_PUBLIC_APP_URL || `http://localhost:${DEV_PORT}`);

if (typeof fetch === 'undefined') {
  console.error('[run-migrations] global fetch is not available. Please run with Node 18+ (fetch builtin) or use a polyfill.');
  process.exit(1);
}

async function runMigrations() {
  console.log('🚀 Running Supabase Migrations');
  console.log(`📍 API URL: ${baseUrl}/api/admin/run-migrations\n`);

  try {
    const response = await fetch(`${baseUrl}/api/admin/run-migrations`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
    });

    const data = await response.json();

    if (!response.ok) {
      console.error('❌ Error:', data.error);
      process.exit(1);
    }

    // Print summary
    console.log('📊 Migration Results:');
    console.log(`   ✓ Executed: ${data.summary?.executed ?? '0'}`);
    console.log(`   ⏭️  Skipped: ${data.summary?.skipped ?? '0'}`);
    console.log(`   ❌ Errors: ${data.summary?.errors ?? '0'}`);
    console.log(`   📈 Total: ${data.summary?.total ?? (Array.isArray(data.results)?data.results.length:'?')}\n`);

    // Print individual results
    if (Array.isArray(data.results) && data.results.length > 0) {
      console.log('📝 Detailed Results:');
      data.results.forEach((result) => {
        const icon =
          result.status === 'success' ? '✓' :
          result.status === 'skipped' ? '⏭️' :
          '❌';

        const message = result.message ? ` (${result.message})` : '';
        console.log(`   ${icon} ${result.file}${message}`);
      });
    }

    if (data.success) {
      console.log('\n🎉 All migrations completed successfully!');
      console.log('✨ Your Supabase database is now ready!');
      process.exit(0);
    } else {
      console.log('\n⚠️  Some migrations failed. Check errors above.');
      process.exit(1);
    }
  } catch (error) {
    console.error('❌ Failed to run migrations:', error instanceof Error ? error.message : String(error));
    console.error('\n💡 Make sure:');
    console.error('   1. Your Next.js app is running (pnpm dev)');
    console.error('   2. The app is accessible at:', baseUrl);
    console.error('   3. Supabase credentials are configured');
    process.exit(1);
  }
}

runMigrations();
